src/osmet/fiemap.rs

```diff
@@ -0,0 +1,173 @@
+// Copyright 2020 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::ffi::OsStr;
+use std::fs::OpenOptions;
+use std::os::unix::io::AsRawFd;
+use std::os::unix::io::RawFd;
+
+use serde::{Deserialize, Serialize};
+
+use crate::errors::*;
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub(super) struct Extent {
+    pub logical: u64,
+    pub physical: u64,
+    pub length: u64,
+}
+
+pub(super) fn fiemap_path(path: &OsStr) -> Result<Vec<Extent>> {
+    let file = OpenOptions::new()
+        .read(true)
+        .open(path)
+        .chain_err(|| format!("opening {:?}", path))?;
+
+    let fd = file.as_raw_fd();
+    Ok(fiemap(fd)?)
+}
+
+
+/// Returns the `Extent`s associated with the given file. Note that the physical offsets are
+/// relative to the partition start on which the file resides.
+fn fiemap(fd: RawFd) -> Result<Vec<Extent>> {
+    let mut m = ffi::fiemap::new();
+    let mut extents: Vec<Extent> = Vec::new();
+
+    loop {
+        m.fm_start = match extents.iter().last() {
+            Some(extent) => extent.logical + extent.length,
+            None => 0,
+        };
+
+        unsafe { ffi::ioctl::fs_ioc_fiemap(fd, &mut m)? };
+        if m.fm_mapped_extents == 0 {
+            break;
+        }
+
+        let mut found_last = false;
+        for extent in m.fm_extents.iter().take(m.fm_mapped_extents as usize) {
+            // XXX: some of these we should handle, others we should pass along higher up
+            if extent.fe_flags & ffi::FIEMAP_EXTENT_NOT_ALIGNED > 0 {
+                return Err("extent not aligned".into());
+            } else if extent.fe_flags & ffi::FIEMAP_EXTENT_MERGED > 0 {
+                return Err("file does not support extents".into());
```

Comment:
It's not strictly an error, but nothing in CoreOS triggers this right now and I'd like to error out for now so that we notice it. Added a comment to clarify this.